File: js/src/strategy/evaluators/InsuranceEvaluatorFactory.js

// Evaluators are depencies for the Factory Module
var SUVInsuranceEvaluator = require("./SUVInsuranceEvaluator");
var InvalidInsuraceEvaluator = require("./InvalidInsuranceEvaluator");

function InsuraceEvaluatorFactory(){
    var that = this;  
    this.evaluators = {
        "SUV": SUVInsuranceEvaluator
    }
    return {
        createEvaluator : function(vehicleType){
          var evaluator = that.evaluators[vehicleType];        
          return evaluator === undefined ? InvalidInsuraceEvaluator : evaluator;
        }  
    };
}

module.exports = InsuraceEvaluatorFactory;

File: java/src/strategy/evaluators/InsuranceEvaluatorFactory.java
package strategy.evaluators;

import strategy.models.vehicles.Vehicle;

public enum InsuranceEvaluatorFactory {

    SUV_VEHICLE("SUV"){
        public InsuranceEvaluator makeInsuranceEvaluator(){
            return new SUVInsuranceEvaluator();
        }
    },
    SEDAN_VEHICLE("Sedan"){
        public InsuranceEvaluator makeInsuranceEvaluator(){
            return new SedanInsuranceEvaluator();
        }
    },
    INVALID_VEHICLE("Default"){
        public InsuranceEvaluator makeInsuranceEvaluator(){
            return new InvalidVehicleEvaluator();
        }
    };

    private String vehicleType;

    InsuranceEvaluatorFactory(String vehicleType){
        this.vehicleType = vehicleType;
    }

    public String getVehicleType(){
        return vehicleType;
    }

    public abstract InsuranceEvaluator makeInsuranceEvaluator();

    public static InsuranceEvaluator createInsuranceEvaluator(Vehicle vehicle){

        try {
            InsuranceEvaluatorFactory evaluator = InsuranceEvaluatorFactory.getVehicle(vehicle.getVehicleType());
            return evaluator.makeInsuranceEvaluator();
        }
        catch ( NullPointerException e){
            return new InvalidVehicleEvaluator();
        }
    }

    public static InsuranceEvaluatorFactory getVehicle(String type){

        for(InsuranceEvaluatorFactory i : InsuranceEvaluatorFactory.values()){

            if( i.getVehicleType().equals(type)){
                return i;
            }
        }
        return InsuranceEvaluatorFactory.INVALID_VEHICLE;
    }
}
